session/src/main/scala/akka/cassandra/session/CqlSessionProvider.scala

```diff
@@ -44,6 +49,75 @@ class DefaultSessionProvider(@unused system: ActorSystem, config: Config) extend
   }
 }
 
+/**
+ * ```
+ * akka {
+ *   discovery.method = config
+ * }
+ * akka.discovery.config.services = {
+ *   cassandra-service = {
+ *     endpoints = [
+ *       {
+ *         host = "127.0.0.1"
+ *         port = 9042
+ *       },
+ *       {
+ *         host = "127.0.0.2"
+ *         port = 9042
+ *       }
+ *     ]
+ *   }
+ * }
+ * my-cassandra-session {
+ *   session-provider = "akka.cassandra.session.AkkaDiscoverySessionProvider"
+ *   session-name = "my-session"
+ *   service {
+ *     name = "cassandra-service"
+ *     lookup-timeout = 1 s
+ *   }
+ * }
+ * ```
+ */
+class AkkaDiscoverySessionProvider(system: ActorSystem, config: Config) extends CqlSessionProvider {
+
+  /**
+   * Use Akka Discovery to read the addresses for `serviceName` within `lookupTimeout`.
+   */
+  private def readNodes(serviceName: String, lookupTimeout: FiniteDuration)(
+      implicit system: ActorSystem): Future[immutable.Seq[String]] = {
+    import system.dispatcher
+    val discovery = Discovery(system).discovery
+    discovery.lookup(serviceName, lookupTimeout).map { resolved =>
+      resolved.addresses.map(addr => addr.host + ":" + addr.port)
+    }
+  }
+
+  /**
+   * Expect a `service` section in Config and use Akka Discovery to read the addresses for `name` within `lookup-timeout`.
+   */
+  private def readNodes(config: Config)(implicit system: ActorSystem): Future[immutable.Seq[String]] =
+    if (config.hasPath("service")) {
+      val serviceName = config.getString("service.name")
+      val lookupTimeout = config.getDuration("service.lookup-timeout").asScala
+      readNodes(serviceName, lookupTimeout)
+    } else throw new IllegalArgumentException(s"config $config does not contain `service` section")
+
+  override def connect()(implicit ec: ExecutionContext): Future[CqlSession] = {
+    val sessionName = config.getString("session-name")
+    require(sessionName != null && sessionName.nonEmpty, s"config needs to set a non-empty `session-name`")
+    val contactPoints = readNodes(config)(system)
+    contactPoints.flatMap { contactPoints =>
+      val overload: DriverConfigLoader =
+        DriverConfigLoader
+          .programmaticBuilder()
+          .withString(DefaultDriverOption.SESSION_NAME, sessionName)
+          .withStringList(DefaultDriverOption.CONTACT_POINTS, contactPoints.asJava)
+          .build()
+      CqlSession.builder().withConfigLoader(overload).buildAsync().toScala
+    }
+  }
```

Comment:
After https://github.com/akka/akka-persistence-cassandra/pull/683 this should be:

```scala
  override def connect()(implicit ec: ExecutionContext): Future[CqlSession] = {
    val driverConfig = CqlSessionProvider.driverConfig(system, config)
    val sessionName =
      if (driverConfig.hasPath("basic.session-name")) driverConfig.getString("basic.session-name")
      else ""
    require(sessionName != null && sessionName.nonEmpty, s"config needs to set a non-empty `basic.session-name`")

    readNodes(config)(system).flatMap { contactPoints =>
      val driverConfigWithContactPoints = ConfigFactory.parseString(s"""
        basic.contact-points = [${contactPoints.mkString("\"", "\", \"", "\"")}]
        """).withFallback(driverConfig)

      val driverConfigLoader = DriverConfigLoaderFromConfig.fromConfig(driverConfigWithContactPoints)
      CqlSession.builder().withConfigLoader(driverConfigLoader).buildAsync().toScala
    }
  }
```